Rust Shannon Decomposition

gadial opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary
Port of the QSD code to Rust. Most of the work was done by @mtreinish; ZXZ code is based on a Python version by @ShellyGarion.


### Details and comments

close #12213.
should be merged only after #14721

mtreinish committed:
[WIP] Port QSD to rust

This commit implements the quantum shannon decomposition algorithm in
rust based on the current Python qs_decomposition function implementation.
This function doesn't support the arbitrary 1q and 2q decomposer
arguments so the Python version of the function is retained and used
while those arguments are supported.

Fixes #12213

gadial committed:
Added ZXZ decomposition; working on random unitaries for 4 qubits but failing on 5

mtreinish committed:
Merge remote-tracking branch 'origin/main' into rust-shannon-decomposition

mtreinish committed:
Fix merge conflicts

This commit fixes merge conflicts as the original qsd implementation in
the first commit is about one month old and the APIs it used on main
have changed slightly in the intervening time.

gadial committed:
Fix failures for 2-qubit matrices by using closest unitary

ShellyGarion commented on crates/circuit/src/circuit_data.rs:
```diff
@@ -2532,6 +2532,95 @@ impl CircuitData {
+    /// Clone an empty CircuitData from a given reference.
+    ///
+    /// The new copy will have the global properties from the provided `CircuitData`.
+    /// The the bit data fields and interners, global phase, etc will be copied to
```
"The the" ?

ShellyGarion commented on crates/circuit/src/circuit_data.rs:
```diff
@@ -2720,6 +2809,10 @@ impl CircuitData {
         .iter()
         .map(|stretch| self.stretches.get(*stretch).unwrap())
     }
+    pub fn reverse(mut self) -> Self {
```
why call it "reverse" and not "reverse_ops" like in the python API? (although I personally this "reverse" is simpler)

gadial replied:
`reverse_ops` creates a copy of the circuit with reversed op, while this method reverses in-place, so I felt we need to use different names. In my opinion `reverse_ops` was not a suitable name due to Python's convention of using `reverse` for the in-place reversing as opposed to `reversed()` which creates an iterator.

ShellyGarion commented on crates/synthesis/src/linalg/cos_sin_decomp.rs:
```diff
@@ -225,7 +225,13 @@ pub fn cos_sin_decomposition(u: DMatrix<Complex64>) -> CosSinDecompReturn {
+    }
 }
 
 // TODO: Remove this function and all the python interface when QSD is ported to rust
```
should this line be removed now?

gadial replied:
I'm waiting for the code to stabilize before removing dead code, but it's very likely this indeed will be removed.

mtreinish replied:
Do we want to remove this now? Or are we still using CSD in python?

ShellyGarion replied:
perhaps this can be done as part of https://github.com/Qiskit/qiskit/pull/14967 ?

ShellyGarion commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,714 @@
+    py: Python,
+    mat: &DMatrix<Complex64>,
+    opt_a1: bool,
+    opt_a2: bool,
```
Note that in #14721 the default was changed to "None" and then we choose the best optimization:
- opt_a1 = True
- opt_a2=False if the unitary is controlled and opt_a2=True otherwise

ShellyGarion commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,714 @@
+    OnlyW,
+}
+
+pub fn quantum_shannon_decomposition(
```
should also copy some of the python documentation

ShellyGarion commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,714 @@
+            Param::Float(sequence.global_phase()),
+        );
+    }
+    // Check whether the matrix is equivalent to a block diagonal w.r.t ctrl_index
```
if opt_a2=None should update opt_a2 accordingly

ShellyGarion commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,714 @@
+/// Given a matrix that is "close" to unitary, returns the closest
+/// unitary matrix.
+/// See https://michaelgoerz.net/notes/finding-the-closest-unitary-for-a-given-matrix/,
+fn closest_unitary(mat: DMatrix<Complex64>) -> DMatrix<Complex64> {
```
this seems to be a useful function also for other parts of the code. should we move it to linalg folder?

ShellyGarion commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,714 @@
+    &u * &v_t
+}
+
+fn _zxz_decomp_svd(A: DMatrix<Complex64>) -> (DMatrix<Complex64>, DMatrix<Complex64>) {
```
perhaps the zxz decomposition should be moved to another file in the linalg folder?

gadial replied:
Does it have uses except for QSD?

ShellyGarion replied:
this is a general decomposition for unitary matrices, although it's not so well-known.
I agree that it may not have any other applications in qiskit except of unitary synthesis.

ShellyGarion commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,714 @@
+    (A1, A2, B, C)
+}
+
+fn demultiplex(
```
should copy the docs from the python code here

ShellyGarion replied:
note that the circuit image doesn't look well

ShellyGarion commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,714 @@
+    Ok((out, vmat, wmat))
+}
+
+// TODO: not sure that's a correct implementation of the python code
```
should we remove this line?

gadial replied:
Already did, because now I'm (relatively) sure :-)

ShellyGarion commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,714 @@
+    }
+    Ok(out)
+}
+fn decompose_uc_rotation(
```
please add some docstring.

ShellyGarion commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,714 @@
+    Ok(out)
+}
+
+fn decompose_uc_rotations(
```
seems that this function appears twice?

ShellyGarion commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,714 @@
+    [(angle_1 + angle_2) / 2., (angle_1 - angle_2) / 2.]
+}
+
+fn get_ucry_cz(num_qubits: u32, angles: &mut [f64]) -> PyResult<CircuitData> {
```
this function is not needed anymore for the ZXZ decomposition

ShellyGarion commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,714 @@
+    Ok(out)
+}
+
+fn apply_a2(
```
need to copy the docs from the python code.

ShellyGarion commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,714 @@
+    Ok(out_circ)
+}
+
+fn append(circ: &mut CircuitData, new: CircuitData, qubit_map: &[Qubit]) -> PyResult<()> {
```
this seems to be a general CircuitData function, why is it located here?

gadial replied:
@mtreinish added it here, so I'm not sure if that's on purpose since it's too hacky in nature, or whether we can move it.

ShellyGarion commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,714 @@
+    Ok(())
+}
+
+fn extract_multiplex_blocks(umat: &DMatrix<Complex64>, k: usize) -> [DMatrix<Complex64>; 4] {
```
need to copy some docs from the python code

ShellyGarion commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,714 @@
+    ]
+}
+
+fn off_diagonals_are_zero(
```
need to copy some docs from the python code

ShellyGarion commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,714 @@
+    py: Python,
+    mat: PyReadonlyArray2<Complex64>,
+    opt_a1: bool,
+    opt_a2: bool,
```
Note that opt_a1 and opt_a2 could be None (after #14721)

ShellyGarion reviewed (commented):
I know that this is a preliminary version of the code, but I still wanted to look at it.
My main comment is that we need to merge #14721 as soon as possible, and then we need to fix the API (note that opt_a1 and opt_a2 could be None) and handle the updated tests.
The current tests do not fit the updated ZXZ algorithm.

gadial committed:
Code cleanup

gadial committed:
Changes according to PR review

ShellyGarion commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,748 @@
+    (s, u)
+}
+
+// fn _zxz_decomp_verify(
```
perhaps the verification that we get back the same matrix could be done using a rust test?
see the suggestion [here](https://github.com/Qiskit/qiskit/pull/14721#discussion_r2243371639).

ShellyGarion replied:
copy of the following comment:
https://github.com/Qiskit/qiskit/pull/14721#discussion_r2245408351

ShellyGarion replied:
done.

ShellyGarion commented on crates/synthesis/src/linalg/mod.rs:
```diff
@@ -10,4 +10,41 @@
+/// Given a matrix that is "close" to unitary, returns the closest
+/// unitary matrix.
+/// See https://michaelgoerz.net/notes/finding-the-closest-unitary-for-a-given-matrix/,
+pub fn closest_unitary(mat: DMatrix<Complex64>) -> DMatrix<Complex64> {
```
it seems to duplicates with https://github.com/Qiskit/qiskit/blob/2faea3849003580c69dc906c2035e743d768efa4/crates/synthesis/src/linalg/cos_sin_decomp.rs#L70
perhaps this function should be used by both cos-sin and zxz ?

mtreinish replied:
Is this still a todo to deduplicate these?

ShellyGarion replied:
the duplication in cosine-sine decomposition is handled as part of https://github.com/Qiskit/qiskit/pull/14967.

gadial committed:
Merge branch 'main' into rust-shannon-decomposition

gadial committed:
A2 optimization fixes

gadial committed:
Fix for axes handling in `extract_multiplex_blocks`

gadial committed:
Linting

ShellyGarion commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,780 @@
+        let orthonormal_eigenvectors = QR::new(eigh.eigenvectors).q();
+        (eigvals, orthonormal_eigenvectors)
+    } else {
+        let shur = nalgebra::linalg::Schur::try_new(um0um1, 1e-12, 100000).unwrap();
```
should be schur and not shur in the variable name

ShellyGarion committed:
add condition number calculation (not actually needed for QSD but may be generally useful)

ShellyGarion committed:
add verification checks for linear algebra methods

mtreinish commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,837 @@
+        return Err(QiskitError::new_err(
+            "The matrix after ZXZ decomposition is not the same.",
+        ));
+    }
```
You can leverage `debug_assert!` here to make this a runtime check only in a debug build: https://doc.rust-lang.org/core/macro.debug_assert.html

This would let us validate by default for rust tests and when we run qiskit when building in unoptimized mode (which we do at least for testing with coverage collection in CI) without incurring the runtime cost for release builds.

alexanderivrii commented:
A few comments after experimenting with this code. I see exactly the same test failures on my mac laptop as in the CI. I can make all of the tests pass, but not in the way that I fully like.

First, the output seems correct, up to numerical errors. The test failures only occur for larger matrices, such as 1-controlled 5-qubit matrices, or 2-controlled 4-qubit matrices. The tests fail because the original matrix and the matrix of the synthesized circuit are not sufficiently close (i.e. not up to `1e-8` in the $L^\infty$-norm, which I think is what `self.assertEqual(op1, op2)` checks). 

I still do not fully understand where numerical precision errors arise and how they propagate from one function to another. Here are some preliminary observations :
* The nalgebra's SVD decomposition may give an imprecise decomposition (failing the new internal rust check `_zxz_decomp_verify`, also note that this check only checks up to `1e-7` in the $L^\infty$-norm, not `1e-8` used in python testing). This can be "fixed" by replacing `a.SVD` by `a.try_svd(true, true, 1e-12, 0)` in the method `_zxz_decomp_svd`.
* When the input matrices to some of the methods (for instance, to `demultiplex`) are not sufficiently unitary (to be precise, when`u.adjoint() * u` is not sufficiently close to the identity matrix), the decompositions produced by the methods are not sufficiently precise either. This is already partially handled in the code by calling `closest_unitary` in several places.
* The method `closest_unitary` does not always do the perfect job, it itself depends on SVD, and again it helps to replace `SVD` by `try_svd` in the code for `closest_unitary`.
* Even then, the `demultiplex` still gets called with non-sufficiently-unitary matrices. As a hack, I have changed my code to modify the expected-to-be-unitary input matrices `um0` and `um1` by closest-unitary representatives. Now all of the internal rust checks are happy (up to `1e-7`).
* For consistency, I have modified the Python tests to also check up to `1e-7` (and because they fail otherwise), now they pass.

My local code involves too much cloning and too many calls to `closest_unitary`, so I am not very happy with that. A better strategy would be to try to understand how the errors propagate in the overall algorithm.

ShellyGarion committed:
replace svd by try_svd

ShellyGarion committed:
update atol

ShellyGarion committed:
merge with main branch

ShellyGarion committed:
remove verify

ShellyGarion committed:
add closest_unitary

ShellyGarion committed:
remove closest_unitary

ShellyGarion committed:
update atol as a function of the number of qubits

ShellyGarion committed:
add move verification methods. add closest_unitary in demultiplex.

ShellyGarion committed:
update tol in tests to 1e-6

ShellyGarion committed:
minor fix

ShellyGarion committed:
merge main branch

ShellyGarion committed:
update EPS in internal verification

ShellyGarion committed:
update EPS in internal verification

ShellyGarion committed:
update EPS in internal verification

ShellyGarion committed:
update tolerance in tests

ShellyGarion committed:
try to make linux-arm tests pass

alexanderivrii commented:
Update: using faer's SVD instead of nalgebra's one significantly improves numerical accuracy (on my arm-based mac laptop) and makes all of the python QSD tests pass with the original error tolerance of `1e-8`. Note that for this experiment I did not rewrite the full algorithm using faer, I am only using faer when calling SVD (that is, convert nalgebra's matrix to faer, calling faer's SVD, and converting the results back to nalgebra). 

For fun: Qiskit-Aer also has an implementation of SVD (written in C++): https://github.com/Qiskit/qiskit-aer/blob/main/src/simulators/matrix_product_state/svd.cpp. However, the code seems significantly less clean and potentially buggy (I have locally ported it to Rust, and noticed some potential problems with the loop bounds).

ShellyGarion committed:
Merge branch 'main' into rust-shannon-decomposition

ShellyGarion committed:
replace nalgebra SVD by faer SVD.
Co-authored-by: Alexander Ivrii <alexi@il.ibm.com>

ShellyGarion committed:
update SVD to faer, updating bounds, formatting

ShellyGarion committed:
update bounds in tests

ShellyGarion committed:
update bounds, formatting

ShellyGarion committed:
update bounds in tests

ShellyGarion committed:
debug_assert in verification functions

ShellyGarion commented:
Finally, all the tests are passing and this PR seems to be ready (but there are still some clippy errors when using nalgenra `stack!`)  - so I decided to ignore them for now.

ShellyGarion committed:
ignore clippy errors in stack!

ShellyGarion committed:
replace zero by 0 in stack

ShellyGarion committed:
fix lint errors

mtreinish commented on crates/circuit/src/circuit_data.rs:
```diff
@@ -2563,6 +2563,95 @@ impl CircuitData {
+    /// Get a mutable view of the instructions in the circuit data
+    pub fn data_mut(&mut self) -> &mut [PackedInstruction] {
+        &mut self.data
+    }
```
I'm pretty sure this was in my original draft commit but we need to be careful adding this, potentially marking it unsafe. Using `data_mut()` requires the users to worry about all the invariants themselves and assert they are doing it correctly otherwise they could corrupt the circuit. In my draft commit I was just being lazy and added this while in practice we are only doing this  is invalidating a label on a single instruction which we should be able to do through a safe api (and if not we should just add a limited rust method for that).

 I might have been thinking there was a use case for this elsewhere in the function but that never materialized. In practice we can't merge this PR while exposing this because it's not a safe interface even if it's internal to rust.

gadial replied:
Removed this, replaced by `invalidate_label` which only removes the label from an instruction of a given index. Note that we suspect the whole code in A2 is not required since this case will never arise. But we need to be sure of it.

mtreinish commented on crates/circuit/src/circuit_data.rs:
```diff
@@ -2751,6 +2840,10 @@ impl CircuitData {
+        self.data.reverse();
+        self
+    }
 }
```
If we're adding a method for this we need to document it, especially since it's an inplace modification.

That being said this method isn't safe, if nothing else it would result in an invalid parameter table because the instruction indices are no longer correct. Realistically this isn't a safe mutation to do in place because the indices are significant and are shared references elsewhere in the circuit. If we need a method to reverse this and just a reverse iterator (which we can do with something like CircuitData::data().iter().rev()`) isn't sufficient we probably need to do:

```suggestion
    pub fn reverse(&self) -> Self {
        let mut out = self.clone_empty_like();
        for inst in self.data().iter().rev() {
            out.push(inst.clone())
        }
        out
    }
}
```

gadial replied:
Used your suggestion (the syntax is slightly different)

mtreinish commented on crates/synthesis/src/linalg/mod.rs:
```diff
@@ -10,4 +10,134 @@
+    let mat_check = v * s * w;
+
+    let close = abs_diff_eq!(mat, &mat_check, epsilon = 1e-7);
+    debug_assert!(close);
```
This isn't doing the `debug_assert!` quite right. You want to put the call to this function inside a `debug_assert!` Right now we're doing the matrix multiplication even in release mode, but not checking the result unless in debug mode.  The goal with using `debug_assert!` here is to avoid the overhead of doing the math to verify the answer is correct in release mode, but leave the checks in debug mode. So we need to makes sure the `debug_assert!` wraps the whole function not just the inner boolean.

```suggestion
```

gadial replied:
Agree, fixed this (also relevant to the next `debug_assert!` comments)

mtreinish commented on crates/synthesis/src/linalg/mod.rs:
```diff
@@ -10,4 +10,134 @@
+fn faer_to_nalgebra(mat: &Mat<Complex64>) -> DMatrix<Complex64> {
+    let (rows, cols) = mat.shape();
+    DMatrix::from_fn(rows, cols, |i, j| mat[(i, j)])
+}
```
If we add `"nalgebra"` to the used features for `faer_ext` here" https://github.com/Qiskit/qiskit/blob/main/crates/synthesis/Cargo.toml#L39 then we can use the built-in conversion traits instead of these functions. The differences with the trait implementations for `.into_nalgebra()` and `.into_faer()` is they work on view objects with no copies. So in some places we will need an explicit `.to_owned()` after doing conversion if we need an owned copy (for example returning the results from svd we want as an owned memory because the faer result object doesn't live long enough for a reference). But we can still avoid at least one copy by using the conversions from faer.

gadial replied:
Done, seems to work well.

mtreinish commented on crates/synthesis/src/linalg/mod.rs:
```diff
@@ -10,4 +10,134 @@
+
+pub fn svd_decomposition_using_faer(
+    mat: &DMatrix<Complex64>,
+) -> (DMatrix<Complex64>, DMatrix<Complex64>, DMatrix<Complex64>) {
```
Why do you need 2 public functions for this?

gadial replied:
Agree, `svd_decomposition_using_faer` should not be public.

mtreinish commented on crates/synthesis/src/linalg/mod.rs:
```diff
@@ -10,4 +10,134 @@
+    mat: &DMatrix<Complex64>,
+) -> (DMatrix<Complex64>, DMatrix<Complex64>, DMatrix<Complex64>) {
+    let faer_mat = nalgebra_to_faer(mat);
+    let faer_svd = faer_mat.svd().expect("Call to Faer failed");
```
I would just use `unwrap()` here so we can at least see the error message from faer.

gadial replied:
Done

mtreinish commented on crates/synthesis/src/linalg/mod.rs:
```diff
@@ -10,4 +10,134 @@
+
+    let u_faer = faer_svd.U().to_owned();
+    let s_faer = faer_svd.S().to_owned();
+    let v_faer = faer_svd.V().adjoint().to_owned();
```
Why does these need to be cloned here? Everything works with references and we copy it during the output conversion to nalgebra. We should only need to copy it once.

gadial replied:
Switched to using views.

mtreinish commented on crates/synthesis/src/linalg/mod.rs:
```diff
@@ -10,4 +10,134 @@
+        } else {
+            Complex64::new(0.0, 0.0)
+        }
+    });
```
Do you need the intermediate faer matrix here. Couldn't this just be an nalgebra matrix `from_fn()`  directly. As it stands now this is copying twice.

gadial replied:
Agree, I generate an nalgebra matrix here now.

mtreinish commented on crates/synthesis/src/linalg/mod.rs:
```diff
@@ -10,4 +10,134 @@
+    let s_na = faer_to_nalgebra(&sigma);
+    let v_na = faer_to_nalgebra(&v_faer);
+
+    verify_svd_decomp(&mat.clone(), &u_na, &s_na, &v_na);
```
```suggestion
    debug_assert!(verify_svd_decomp(&mat.clone(), &u_na, &s_na, &v_na));
```

mtreinish commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,832 @@
+///            `arXiv:0406176 [quant-ph] <https://arxiv.org/abs/quant-ph/0406176>`_
+///         2. Krol, Al-Ars, *Beyond Quantum Shannon: Circuit Construction for General
+///            n-Qubit Gates Based on Block ZXZ-Decomposition*,
+///            `arXiv:2403.13692 <https://arxiv.org/abs/2403.13692>`_
```
This docstring is overindented it just needs one space between `///` and the start of the line. Also this is in the wrong syntax. This is a python Sphinx/rst/napolean docstring but this is a rust function which uses markdown/[rustdoc](https://doc.rust-lang.org/rustdoc/index.html). This should probably move to the `pyfunction` and use markdown for this function

gadial replied:
I fixed this, but I'm not sure about the need for a conversion to markdown. Since I move it to the `pyfunction`, won't it still be compiled and displayed by the sphinx engine?

mtreinish commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,832 @@
+    let default_2q_decomposer = TwoQubitBasisDecomposer::new_inner(
+        StandardGate::CX.into(),
+        smallvec![],
+        StandardGate::CX.matrix(&[]).unwrap().view(),
```
We probably should just use `aview2(&qiskit_circuit::gate_matrix::CX_GATE)` here to save the unnecessary allocation.

gadial replied:
Done

mtreinish commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,832 @@
+///            n-Qubit Gates Based on Block ZXZ-Decomposition*,
+///            `arXiv:2403.13692 <https://arxiv.org/abs/2403.13692>`_
+pub fn quantum_shannon_decomposition(
+    mat: &DMatrix<Complex64>,
```
I'm thinking this might be cleaner pivoting to ndarray `ArrayView2` here instead of `DMatrix` and using `ndarray` and faer` instead of nalgebra. The use of nalgebra here was originally for 2 reasons, the first is csd was written using nalgebra and we were originally using csd here. But the second reason is that we were looking at removing faer as a dependency (which was the motivation for using nalgebra in csd too), but I don't think that's realistic at this point since we are relying on it for svd here. Also nalgebra works best in cases where we have fixed size small matrices but for qsd we don't have that. 

Using ndarray and faer is a more natural fit for how qsd works. It also is a cleaner fit to the use of the 1q and 2q decomposers which are taking an `ArrayView2` as their inputs right now so we'd have simpler code with less conversions between the 3 libraries.

This all assume that's faer has all the functions we need here. I think it should have all the linalg functions we need, but if it's missing something and there isn't a straightforward alternative then we can keep using nalgebra.

gadial replied:
Since this is a major change (unless I'm missing something) I suggest we delay it to a future PR.

mtreinish replied:
I'm fine deferring it, it is a large change to the PR. I think the current state of the PR is a bit better than before using proper view types and the faer-ext crate to do the conversion. That being said this might not hold true anymore if we finish #13665 for the two qubit decomposer module. If we move to using nalgebra types for the input and outputs of the two qubit decomposers using ndarray here for the interface types might not be the best choice anymore anyway, so it's probably better to wait here.

mtreinish commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,832 @@
+    // perform block ZXZ decomposition from [2]
+    let (a1, a2, b, c) = _block_zxz_decomp(mat);
+
+    _zxz_decomp_verify(mat, &a1, &a2, &b, &c);
```
```suggestion
    debug_assert!(_zxz_decomp_verify(mat, &a1, &a2, &b, &c));
```

mtreinish commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,832 @@
+    DMatrix<Complex64>,
+    DMatrix<Complex64>,
+) {
+    verify_unitary(mat);
```
```suggestion
    debug_assert!(verify_unitary(mat));
```

mtreinish commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,832 @@
+
+    const EPS: f64 = 1e-7;
+    let close = abs_diff_eq!(mat, &mat_check, epsilon = EPS);
+    debug_assert!(close);
```
```suggestion
```

mtreinish commented on crates/synthesis/src/linalg/mod.rs:
```diff
@@ -10,4 +10,134 @@
+    let uu = u.adjoint() * u;
+
+    let close = abs_diff_eq!(uu, id_mat, epsilon = 1e-7);
+    debug_assert!(close);
```
```suggestion
```

mtreinish commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,832 @@
+    vw_type: VWType,
+) -> PyResult<(CircuitData, DMatrix<Complex64>, DMatrix<Complex64>)> {
+    verify_unitary(um0);
+    verify_unitary(um1);
```
Do we need these checks if we're calling `closest_unitary()` below?

gadial replied:
Removed.

mtreinish commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,832 @@
+    let um1 = closest_unitary(um1.clone());
+
+    verify_unitary(&um0);
+    verify_unitary(&um1);
```
Similarly shouldn't these be unitary after calling `closest_unitary`?

alexanderivrii replied:
I am also wondering whether these and other calls to `closest_unitary` can now be completely removed?

mtreinish replied:
I was thinking the same thing. If we can drop them that would be best I think

ShellyGarion replied:
these calls to `closest_unitary` in demultiplex function are needed, since otherwise um0^{dg} isn't the same as um0^{inv}, but the verification `verify_unitary(&um0);` could be removed (and was removed).

gadial replied:
All other calls to `closest_unitary` were removed and it seems all the CI tests are passing.

mtreinish commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,832 @@
+            .get_qargs(inst.qubits)
+            .iter()
+            .map(|x| qubit_map[x.index()])
+            .collect();
```
I think there is a path to avoid the allocation here. @jakelishman might have a pattern off the top of his head, but allocating a vec on each iteration seems unnecessary for this.

gadial replied:
I guess we can always use a dict to store pre-encountered mapped qubits:

```
    let mut new_qubits_map: HashMap<Interned<[Qubit]>, Vec<Qubit>> = HashMap::new();
    for inst in new.data() {
        let qubits_map = new_qubits_map.entry(inst.qubits).or_insert_with(|| { 
            new
            .get_qargs(inst.qubits)
            .iter()
            .map(|x| qubit_map[x.index()])
            .collect()
        });
```

Do you think this can be further optimized?

mtreinish commented on qiskit/synthesis/unitary/qsd.py:
```diff
@@ -106,6 +107,13 @@ def qs_decomposition(
+        return QuantumCircuit._from_circuit_data(qsd.qs_decomposition(array, opt_a1, opt_a2))
+
+    # TODO call rust if decomposer_1q or decomposer_2q are OneQubitEulerDecomposer or
+    #  TwoQubitBasisDecomposer, deprecate otherwise
```
We should do this still we can save the deprecation for a follow on, but if we believe there is value in allowing users to provide their own 1q and 2q decomposer for use internally we should pass the objects to the inner rust function if their of the correct type.

gadial replied:
Is this actionable in this PR or should we wait for a future update?

gadial replied:
Added the check and removed the comment; we should deprecate in a separate PR.

mtreinish commented on test/python/synthesis/test_qsd_synthesis.py:
```diff
@@ -355,7 +359,7 @@ def test_mc_1qubit_opt(self, num_qubits, base_gate):
         cqc2 = transpile(qc2, basis_gates=["u", "cx"], optimization_level=0)
         op2 = Operator(qc2)
-        self.assertEqual(hidden_op, op2)
+        self.assertTrue(matrix_equal(hidden_op.to_matrix(), op2.to_matrix(), atol=1e-8))
```
What was the tolerance on this check before moving to rust? I'm wondering how loose we're making this now compared to where we were before.

ShellyGarion replied:
the default tolerance is 1e-8:
https://github.com/Qiskit/qiskit/blob/4bf1e26c319c899b2b925a8b80231d880263ce83/qiskit/quantum_info/operators/predicates.py#L20

mtreinish reviewed (changes requested):
It's great to see this finally all working, it's been a couple of months since I wrote up the first commit here to get this started so I'm really glad we've almost at the point where we can use this inside the transpiler for unitary synthesis from rust without Python. This also gives us a path to doing 3q unitary peephole now that we have the block zxz decomposition too.

That being said I left a bunch of inline comments I think we need to adjust some of the internals here as there are some sharp edges to the current implementation. I'm responsible for some of that as in my early WIP draft commit I was playing fast and loose with things to get something working fast. But before we can move forward with this we'll need to clean these this up.

gadial committed:
Changes according to PR review

gadial committed:
Merge branch 'main' into rust-shannon-decomposition

gadial committed:
Switch to default faer-nalgebra conversions, Partially return closest_unitary

gadial committed:
Return all closest_unitary and linting

gadial committed:
Fixes in `svd_decomposition_using_faer`

gadial committed:
Memory optimizations and removal of one `closest_unitary` usage

gadial committed:
Remove one use of `closest_unitary` from A2

ShellyGarion commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,813 @@
+
+    const EPS: f64 = 1e-7;
+    let close = abs_diff_eq!(u_block, u_check, epsilon = EPS);
+    debug_assert!(close);
```
should we leave the debug_assert here?

gadial replied:
No, I missed this one. Will fix.

ShellyGarion commented on crates/synthesis/src/linalg/mod.rs:
```diff
@@ -10,4 +10,123 @@
+const ATOL_DEFAULT: f64 = 1e-8;
+const RTOL_DEFAULT: f64 = 1e-5;
+
+pub fn is_hermitian_matrix(mat: &DMatrix<Complex64>) -> bool {
```
I think that it's better that all functions in this file (which could be useful for other applications) would have some docstring.

gadial replied:
Added

ShellyGarion commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,813 @@
+    let um1 = closest_unitary(um1.clone());
+
+    debug_assert!(verify_unitary(&um0));
+    debug_assert!(verify_unitary(&um1));
```
if we have to call `closest_unitary` here, maybe we don't need `debug_assert!` ?

gadial replied:
I thought we also wanted to double-check `closest_unitary` but I guess it's not critical. I'll remove this.

ShellyGarion commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,813 @@
+    let d_mat: DMatrix<Complex64> = DMatrix::from_diagonal(&d_values);
+    let wmat = d_mat.clone() * vmat.adjoint() * um1.clone();
+
+    _demultiplex_verify(&um0, &um1, &vmat, &wmat, &d_mat);
```
here maybe we should have `debug_assert!`

gadial replied:
Added

ShellyGarion commented on test/python/synthesis/test_qsd_synthesis.py:
```diff
@@ -338,9 +339,12 @@ def test_block_diag_opt(self, num_qubits):
+    )
     def test_mc_1qubit_opt(self, num_qubits, base_gate):
-        """Create a multi-controlled X, P or U gate on num_qubits.
+        """Create a multi-controlled Z, P or U gate on num_qubits.
```
here it should be X, Z, P or U

gadial replied:
Done

ShellyGarion reviewed (commented):
Thanks @gadial for handling the review comments well. I only have a few minor comments.

gadial committed:
Remove the second usage of `closest_unitary` in A2

gadial committed:
Additional small fixes according to PR review

alexanderivrii commented on crates/synthesis/src/linalg/cos_sin_decomp.rs:
```diff
@@ -23,13 +23,13 @@ use num_complex::{Complex64, ComplexFloat};
+    pub r0: DMatrix<Complex64>,
+    pub r1: DMatrix<Complex64>,
+    pub thetas: Vec<f64>,
+}
```
While I agree that this makes the return structure more clear, why did you need to make changes to this file? I think it would make sense to add them to #14967 instead.

ShellyGarion replied:
I think these changes were done in older versions of the code which was based on cosine-sine decomposition and not on ZXZ. I agree that maybe we don't need them as part of this PR.

alexanderivrii commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,800 @@
+    let u = v * w_dg;
+
+    (s, u)
+}
```
I have two small comments about this (which also apply to some other functions in this and the mod.rs file). First, I don't think we prefix function names in the Rust code with the underscore. Second, I believe we should change `a` to be of type `DMatrixView`, which would avoid the need to clone matrices when calling this function from other places.

alexanderivrii commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,800 @@
+    let u21 = mat.view((n, 0), (n, n));
+    let u22 = mat.view((n, n), (n, n));
+    let (sx, ux) = _zxz_decomp_svd(x.into_owned());
+    let (sy, uy) = _zxz_decomp_svd(y.into_owned());
```
The `into_owned` can be removed if `_zxz_decomp_svd` works with `DMatrixView` (and note that `x` and `y` are already views).

alexanderivrii commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,800 @@
+    let a_block = stack![a1, 0; 0, a2];
+
+    let b1 = iden.clone() + b;
+    let b2 = iden.clone() - b;
```
This code will only be used as a part of self-sanity check in the non-release compilation but anyhow the cloning here can be removed:
```suggestion
    let b1 = &iden + b;
    let b2 = &iden - b;
```

alexanderivrii commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,800 @@
+    vw_type: VWType,
+) -> PyResult<(CircuitData, DMatrix<Complex64>, DMatrix<Complex64>)> {
+    let um0 = closest_unitary(um0.clone());
+    let um1 = closest_unitary(um1.clone());
```
You can avoid cloning the matrices if `closest_unitary` works with `DMatrixView` instead.

alexanderivrii commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,800 @@
+        let orthonormal_eigenvectors = QR::new(eigh.eigenvectors).q();
+        (eigvals, orthonormal_eigenvectors)
+    } else {
+        let schur = nalgebra::linalg::Schur::try_new(um0um1, 1e-12, 100000).unwrap();
+        let (vmat, evals) = schur.unpack();
+        let eigvals = evals.diagonal();
+        (eigvals, vmat)
```
Do you understand why this works out-of-the-box here and not with faer? As you have explained to me yesterday, Schur's decomposition (https://en.wikipedia.org/wiki/Schur_decomposition) expresses $A = Q U Q^{-1}$, where $U$ is unitary. For unitary matrices, $Q^{-1}$ is the same as $Q^*$, allowing to consider `vmat.adjoint()` in the code below. However, as we saw with faer, due to the presence of numerical errors, in practice the matrix $Q$ (here `vmat`) might be close but not exactly unitary, in which case we either need to work with `vmat.inverse` instead of `vmat.adjoint` or (better yet) run `close_unitary` on `vmat`.  This seems crucial for faer, but not here, why?

gadial replied:
I don't know. My only guess is that the schur code gives a good approximation.

alexanderivrii commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,800 @@
+        .chain([_ctrl_index])
+        .map(Qubit::new)
+        .collect();
+    let um0um1 = um0.clone() * um1.adjoint();
```
You can avoid cloning by writing (not sure how to do this in 1 line):
```
let um0_view: DMatrixView<Complex64> = um0.as_view();
let um0um1 = um0_view * um1.adjoint();
```

gadial replied:
We can do it as a one-liner by doing
```
let um0um1 = &um0 * um1.adjoint();
```

The `adjoint()` still creates a copy; we may be able to optimize this away as well, but I don't know how - nalgebra supports [ad_mul](https://docs.rs/nalgebra/latest/nalgebra/base/struct.Matrix.html#method.ad_mul) which does `self.adjoint() * rhs`, which is the opposite of what we need. Really strange there's nothing for the other case.

alexanderivrii replied:
Oh interesting, with `faer` the adjoint seems to give a view (a `MatRef`) and not an owned matrix.

alexanderivrii commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,800 @@
+    };
+    let d_values: DVector<Complex64> = eigvals.map(|x| x.sqrt());
+    let d_mat: DMatrix<Complex64> = DMatrix::from_diagonal(&d_values);
+    let wmat = d_mat.clone() * vmat.adjoint() * um1.clone();
```
You can avoid cloning here
```suggestion
    let wmat = &d_mat * vmat.adjoint() * um1;
```

alexanderivrii commented on crates/synthesis/src/linalg/mod.rs:
```diff
@@ -10,4 +10,127 @@
+const RTOL_DEFAULT: f64 = 1e-5;
+
+/// Check whether the given matrix is hermitian by comparing it (up to tolerance) with its hermitian adjoint.
+pub fn is_hermitian_matrix(mat: &DMatrix<Complex64>) -> bool {
```
I don't know if it matters that much, but you can pass `DMatrixView` instead of `&DMatrix`. I think the difference is who owns the underlying memory, and using `DMatrixView` makes the least assumptions.

alexanderivrii commented on crates/synthesis/src/linalg/mod.rs:
```diff
@@ -10,4 +10,127 @@
+/// Given a matrix that is "close" to unitary, returns the closest
+/// unitary matrix.
+/// See https://michaelgoerz.net/notes/finding-the-closest-unitary-for-a-given-matrix/,
+pub fn closest_unitary(mat: DMatrix<Complex64>) -> DMatrix<Complex64> {
```
As mentioned before, you want to work with `DMatrixView` so that you don't need to clone from the calling code.

alexanderivrii commented on crates/synthesis/src/linalg/mod.rs:
```diff
@@ -10,4 +10,127 @@
+    mat: &DMatrix<Complex64>,
+) -> (DMatrix<Complex64>, DMatrix<Complex64>, DMatrix<Complex64>) {
+    svd_decomposition_using_faer(mat)
+}
```
Again we can let the input be `DMatrixView`. I am wondering if we should return just the diagonal (produced by SVD) instead of passing an explicitly created matrix. This is especially true when used for `closest_unitary` since the middle part is ignored anyway.

gadial replied:
I agree, we may be able to optimize here as well. I still need to play with the exact syntax, so for now I'm pushing the rest of the fixes without it.

alexanderivrii reviewed (commented):
This is not a complete review. I have left some comments about switching to `DMatrixView` in a couple of places, which would avoid many unnecessary calls to `clone` and `to_owned`.

Other than that, I have locally replaced all of the code to call faer instead of nalgebra, and experimented (in release compilation of course) with qsd decompositions of random matrices of various size (1000 matrices each). For completeness:
```
import time
from qiskit.quantum_info import random_unitary
from qiskit.synthesis.unitary import qsd

for nq in [2, 3, 4, 5, 6]:
  total_time = 0
  for seed in range(100):
    unitary = random_unitary(2 ** nq, seed=seed)
    time_start = time.time()
    qsd.qs_decomposition(unitary)
    time_end = time.time()
    total_time += (time_end - time_start)
  print(f"nq = {nq}, total_time = {total_time:.4f}")
```
Update: there were some problems with the script above; I need to return the experiments.

We can see that the two sets of runtimes are pretty similar, with your current nalgebra implementation (involving extra clones) doing slightly better than the faer one. I am not sure there is a reason to use faer over nalgebra (except decreasing the number of external crates we rely upon). What do you think?

gadial committed:
Fixes according to PR review

gadial commented:
Thanks @alexanderivrii for your helpful comments. I've fixed every immediate item, which leaves us two:
1. Avoid using `adjoint` and instead use some clever method which multiplies by adjoint without additional allocation (maybe use `ad_mul` somehow?)
2. Change our SVD to return `nalgebra` style diagonal (or maybe even a `Vec`?) and fix the only place in the code where it is used (is there a nice syntax for multiplying an nalgebra matrix by a diagonal matrix given by `Vec` or should we simply iterate over all the cells?

gadial committed:
Call rust for default decomposers

ShellyGarion commented:
Jointly with @alexanderivrii we compared the performance of this PR (rust code) to the main branch (python code) for 100 random matrices of difference sizes.

Shelly's Linux machine:

| num_qubits | main (python) | PR (rust) | ratio | 
|--------------|-----------------|-----------|------|
| 2 | 0.0491 | 0.0206 | x2.3 |
| 3 | 0.9462 | 0.0780 |  x12.1 |
| 4 | 5.1721 | 0.3111 | x16.6 |
| 5 | 28.9451 | 1.3792 | x21 |
| 6 |  162.9357 | 7.6077 | x21.4 |

Sasha's mac laptop:

| num_qubits | main (python) | PR (rust) | ratio | 
|--------------|-----------------|-----------|------|
| 2 | 0.0081| 0.0041 | x1.9  |
| 3 | 0.3829 | 0.0251 |  x15.2  |
| 4 | 1.5683 | 0.1185 |  x13.2 |
| 5 | 9.7425 | 0.5379 | x18.1 |
| 6 |  58.8798 | 2.5207 |  x23.3 |
| 7 |  297.9815 | 12.6091 |  x23.6 |

ShellyGarion commented:
Another performance benchmark of this PR (rust code) to the main branch (python code) for 100 random **controlled** unitary matrices of difference sizes.

Shelly's Linux machine:

| num qubits | num controls | main (python) | PR (rust) | ratio | 
|--------------|----------------|----------------|-----------|-------|
| 2 | 1 | 0.2059 |  0.0371 | x5.5 |
| 2 | 2 | 1.4747 | 0.1389 | x10.6 |
| 2 | 3 | 9.1558 | 0.5611 | x16.3 |
| 3 | 1 | 1.6007 | 0.1262 | x12.6 |
| 3 | 2 | 9.5334 | 0.5825 | x16.3 | 
| 3 | 3 | 58.0699 | 2.6968 | x21.5 | 
| 4 | 1 | 11.5412 | 0.5896 | x19.5 |
| 4 | 2 | 57.2717 | 2.7921 | x20.5 |
| 4 | 3 | 327.4981 | 13.8779 | x23.5 |

mtreinish commented on crates/circuit/src/circuit_data.rs:
```diff
@@ -2563,6 +2563,95 @@ impl CircuitData {
         self.identifier_info.values()
     }
 
+    /// removes the label for the instruction given by index
```
```suggestion
    /// Remove the label for an instruction in the circuit
    ///
    /// This modifies the circuit in place and sets the label
    /// field of an instruction to ``None``.
    ///
    /// #Arguments
    ///
    /// * index: The index of the instruction in the circuit to remove the label of.
```

mtreinish commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,800 @@
+/// the same as the unoptimized method in [1]:
+///
+///     (9/16)4^n - (3/2)2^n
+///
```
Should this be [math mode](https://documentation.help/Sphinx/math.html)? The indent here is a bit odd in sphinx rst. It's not a code block (no `::`) and there isn't a specified directive

gadial replied:
Everything in this docstring was based on the false assumption that this docstring should be in markdown and not sphinx, and unfortunately cargo doesn't support math mode in markdown out of the box (should be bother trying to find an addon?). I'm fixing all these comments by reverting to the sphinx formatting here.

mtreinish replied:
This docstring uses rst and sphinx because it's a pyfunction. The other pure rust functions, that are not exposed to python, use marksown/rustdoc.

gadial replied:
Yeah, this part I understood, but I misunderstood your comment above as saying I should *both* move to the pyfunction *and* change from rst to markdown. Should be ok now.

mtreinish commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,800 @@
+///
+///
+///     (2/3)(4^(n-2) - 1)
+///
```
Same comment here on math mode.

mtreinish commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,800 @@
+///
+///
+///     4^(n-2) - 1.
+///
```
Math mode?

mtreinish commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,800 @@
+///
+///     (22/48)4^n - (3/2)2^n + (5/3)
+///
+///
```
Math mode?

mtreinish commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,800 @@
+    let dim = mat.shape().0;
+    let num_qubits = dim.ilog2() as usize;
+
+    if abs_diff_eq!(DMatrix::identity(dim, dim), mat) {
+        let out_qubits = (0..num_qubits)
+            .map(|_| ShareableQubit::new_anonymous())
+            .collect::<Vec<_>>();
+        return CircuitData::new(Some(out_qubits), None, None, 0, Param::Float(0.));
+    }
```
Should we do this in the inner method instead? I guess it's probably pretty unlikely for one of the recursive calls to be an identity matrix so it's probably not worth the repeated equality check overhead on each recursion.

mtreinish commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,800 @@
+    if dim == 2 {
+        let array = Array2::from_shape_fn((dim, dim), |(i, j)| mat[(i, j)]);
+        let sequence =
+            unitary_to_gate_sequence_inner(array.view(), one_qubit_decomposer, 0, None, true, None);
```
You can avoid the copy here with:

```suggestion
        let dim = ::ndarray::Dim(mat.shape());
        let strides = ::ndarray::Dim(mat.strides());
        // SAFETY: We know the array is a 2x2 and contiguous block (DMatrix uses a vec for backing storage) so we
        // don't need to check for invalid format
        let array = unsafe { ArrayView2::from_shape_ptr(dim.strides(strides), mat.get_unchecked(0)) };
        let sequence =
            unitary_to_gate_sequence_inner(array, one_qubit_decomposer, 0, None, true, None);
```

mtreinish commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,800 @@
+            out.push(packed_inst)?;
+            return Ok(out);
+        }
+        let array = Array2::from_shape_fn((dim, dim), |(i, j)| mat[(i, j)]);
```
Same comment here as above for the unsafe block that can build the array view directly to the DMatrix memory.

mtreinish commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,800 @@
+}
+
+/// Block ZXZ decomposition method, by Krol and Al-Ars [2]
+fn _block_zxz_decomp(
```
We don't need the leading underscore in rust the function is already private.

```suggestion
fn block_zxz_decomp(
```

mtreinish commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,800 @@
+
+/// Verify ZXZ decomposition gives the same unitary
+#[allow(clippy::toplevel_ref_arg)]
+fn _zxz_decomp_verify(
```
Same comment here about the leading underscore.

```suggestion
fn zxz_decomp_verify(
```

mtreinish commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,800 @@
+}
+
+#[allow(clippy::toplevel_ref_arg)]
+fn _demultiplex_verify(
```
```suggestion
fn demultiplex_verify(
```

mtreinish commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,800 @@
+    let u_check = v_block * d_block * w_block;
+
+    const EPS: f64 = 1e-7;
+    abs_diff_eq!(u_block, u_check, epsilon = EPS)
```
Defining a constant like this right before it's only use is a bit odd. It doesn't hurt anything but it's just using two lines for what is identical to:

```suggestion
    abs_diff_eq!(u_block, u_check, epsilon = 1e-7)
```

If we were reusing `EPS` as 1e-7 elsewhere in the file defining a shared `const` value here would be a good choice. Although this shadows an existing `EPS` constant value of 1e-10 which is also not ideal.

mtreinish commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,800 @@
+            // Is this a correct translation from python?
+            // binary_rep = np.binary_repr(i + 1)
+            // q_contr_index = len(binary_rep) - len(binary_rep.rstrip("0"))
+            let q_ctrl_index = (i + 1).trailing_zeros();
```
Is this still an open question? It looks correct to me, the numpy code is trying to compute the number of trailing zeroes which is what the rust method is doing.

gadial replied:
This is a `TODO` comment for me which should have been marked with "TODO". It's usually something I explicitly verify with debug prints, but should have already been removed at this stage.

mtreinish commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,800 @@
+                .iter()
+                .map(|x| qubit_map[x.index()])
+                .collect()
+        });
```
I feel like there is a better way to do this that doesn't involve so many allocations. But I need to think about how to reuse and map the interner indices more directly and check the api for the interners to figure out the correct method to do this so I don't have a concrete suggestion, just to point out this is inefficient.

A small incremental improvement here is to at least do
```suggestion
    let mut new_qubits_map: HashMap<Interned<[Qubit]>, Vec<Qubit>> = HashMap::with_capacity(new.qargs_interner().len());
    for inst in new.data() {
        let qubits_map = new_qubits_map.entry(inst.qubits).or_insert_with(|| {
            new.get_qargs(inst.qubits)
                .iter()
                .map(|x| qubit_map[x.index()])
                .collect()
        });
```
to avoid needing to grow the hashmap repeatedly.

gadial replied:
Am I incorrect in thinking we have one allocation per (mapped) qubit vector used in the instructions, which we must create anyway in order to add to the circuit?

mtreinish replied:
The instructions stored in the circuit used interned indices and represent the qubit (and clbit) arguments as a single u32 internally, which is typed as `Interned<[Qubit]>`. That's what `get_qargs` takes in. The issue is that the interner  (the data structure that stores the mapping of qarga to interned value) and by extension the interned value are tied to each individual circuit and not portable between circuits. So you need to add  the interner mappings from `new` here with the caller supplied `qubit_map` applied to `self``s interner. In my head there should be a method on the interners that does that bulk add and then you are just mapping old interned value to the new interned value.

gadial replied:
So I guess we leave this as-is for now?

mtreinish replied:
I took care of this in: https://github.com/Qiskit/qiskit/pull/14797/commits/7a80f6c1822c01194fe90847afde328acef6a900

mtreinish commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,800 @@
+            let OperationRef::Unitary(unitary) = circ.data()[*idx].op.view() else {
+                unreachable!("diagonal unitary is not a unitary gate");
+            };
+            (*idx, unitary.matrix_view().to_owned())
```
You can just do:

```suggestion
            (*idx, unitary.matrix().unwrap())
```

`.matrix_view()` is to get a no-copy read-only ndarray view of the unitary, while we are explicitly copying here. We don't actually need owned matrices for this outer initialization, but I don't think the matrices generated by `two_qubit_decompose_up_to_diagonal` will live long enough to use a view, so it's fine to use owned here I guess.

gadial replied:
Note that `matrix()` accepts `_params: &[Param]`; I passed `&[]`, hope it's ok.

mtreinish commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,800 @@
+                qubits.iter().map(|q| Qubit(*q as u32)).collect(),
+                vec![],
+            ))
+        }),
```
```suggestion
        qc3_seq.gates().into_iter().map(|(gate, params, qubits)| {
            Ok((
                gate,
                params.into_iter().map(|x| Param::Float(x)).collect(),
                qubits.into_iter().map(|q| Qubit(q as u32)).collect(),
                vec![],
            ))
        }),
```

gadial replied:
This won't work as-is since `qc3_seq.gates()` returns `&TwoQubitSequenceVec`. I tried fixing it by adding

```
pub fn into_gates(self) -> TwoQubitSequenceVec {
        self.gates
    }
```

I hope that's the intended solution and naming convention.

mtreinish reviewed (commented):
The other thing I forgot to mention in my earlier review is I think we could add a synthesis feature release note for this given the performance improvement. If it's that much faster it's worth calling out that qsd for general unitary synthesis is significantly faster for 2.3.

gadial committed:
Fixes according to PR review

gadial commented:
With regards to the comment about https://github.com/Qiskit/qiskit/issues/13665, I'm now giving it another go (it compiles, but even after the initial bugfix round it still gives wrong results). If I succeed, we'll still need an additional PR to convert as much of the rest of the code as possible, and I'll handle this part here as well.

mtreinish commented on crates/circuit/src/circuit_data.rs:
```diff
@@ -2563,7 +2563,14 @@ impl CircuitData {
+    /// This modifies the circuit in place and sets the label
+    /// field of an instruction to ``None``.
+    ///
+    /// #Arguments
```
```suggestion
    /// # Arguments
```

mtreinish commented on crates/synthesis/src/qsd.rs:
```diff
@@ -43,6 +45,18 @@ enum VWType {
+/// Decomposes a unitary matrix into one and two qubit gates using Quantum Shannon Decomposition,
+/// based on the Block ZXZ-Decomposition.
+///
+/// #Arguments
```
```suggestion
/// # Arguments
```

mtreinish commented on crates/synthesis/src/qsd.rs:
```diff
@@ -43,6 +45,18 @@ enum VWType {
+/// * two_qubit_decomposer: Optional alternative two qubit decomposer
+/// * one_qubit_decomposer: Optional alternative one qubit decomposer
+///
+/// Returns:
```
```suggestion
/// # Returns
```

mtreinish commented on crates/synthesis/src/qsd.rs:
```diff
@@ -43,6 +45,18 @@ enum VWType {
+/// * one_qubit_decomposer: Optional alternative one qubit decomposer
+///
+/// Returns:
+///     QuantumCircuit: Decomposed quantum circuit.
```
```suggestion
///     Decomposed quantum circuit.
```

mtreinish commented on crates/synthesis/src/qsd.rs:
```diff
@@ -43,6 +45,18 @@ enum VWType {
+/// * opt_a1: whether to try optimization A.1
+/// * opt_a2: whether to try optimization A.2
+/// * two_qubit_decomposer: Optional alternative two qubit decomposer
+/// * one_qubit_decomposer: Optional alternative one qubit decomposer
```
```suggestion
/// * `mat`: unitary matrix to decompose
/// * `opt_a1`: whether to try optimization A.1
/// * `opt_a2`: whether to try optimization A.2
/// * `two_qubit_decomposer`: Optional alternative two qubit decomposer, if not specified a decomposer using CX and U is used.
/// * `one_qubit_decomposer`: Optional alternative one qubit euler basis to use for single qubit unitary decompositions. If not specified U is used.
```

mtreinish committed:
Optimize interner usage for append() function

This commit adds a two new methods to expose the ability for merging
interners between circuits. This is then used in the qsd code's append
code to simplify the logic and avoid intermediate allocations for each
instruction being moved from one circuit to another.

mtreinish committed:
Merge remote-tracking branch 'origin/main' into rust-shannon-decomposition

mtreinish commented on crates/synthesis/src/linalg/mod.rs:
```diff
@@ -10,4 +10,127 @@
+
+fn svd_decomposition_using_faer(
+    mat: DMatrixView<Complex64>,
+) -> (DMatrix<Complex64>, DMatrix<Complex64>, DMatrix<Complex64>) {
```
```suggestion
/// Returns the SVD decomposition of the given matrix M as three matrices A,S,B such that
/// M=ASB (In the usual notations, M=USV*, and this function returns A=U and B=V*)
pub fn svd_decomposition(
    mat: DMatrixView<Complex64>,
) -> (DMatrix<Complex64>, DMatrix<Complex64>, DMatrix<Complex64>) {
```

mtreinish commented on crates/synthesis/src/linalg/mod.rs:
```diff
@@ -10,4 +10,127 @@
+
+    let max_sv = singular_values
+        .iter()
+        .cloned()
```
```suggestion
        .copied()
```

mtreinish commented on crates/synthesis/src/linalg/mod.rs:
```diff
@@ -10,4 +10,127 @@
+        .fold(f64::NEG_INFINITY, f64::max);
+    let min_sv = singular_values
+        .iter()
+        .cloned()
```
```suggestion
        .copied()
```

mtreinish commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,818 @@
+        1.0,
+        "U",
+        None,
+    )?;
```
Realistically we should be avoiding this on every recursive iteration here. There is a non-zero overhead to construct the two qubit decomposer since it does the weyl decomposition on cx every time. That's wasted work we only need to do it once as it's the same every time.

gadial replied:
Addressed in https://github.com/Qiskit/qiskit/pull/14797/commits/ce20bfbff428095042f62cea5bc9a0c40db3d6e1

mtreinish commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,818 @@
+                ))
+            }),
+            Param::Float(sequence.global_phase()),
+        );
```
```suggestion
        let global_phase = sequence.global_phase();
        return CircuitData::from_packed_operations(
            num_qubits as u32,
            0,
            sequence.into_gates().into_iter().map(|(op, params, qubits)| {
                Ok((
                    op,
                    params.into_iter().map(Param::Float).collect(),
                    qubits.into_iter().map(|q| Qubit(q as u32)).collect(),
                    vec![],
                ))
            }),
            Param::Float(global_phase),
        );
```

mtreinish commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,818 @@
+    let out_qubits = (0..num_qubits)
+        .map(|_| ShareableQubit::new_anonymous())
+        .collect::<Vec<_>>();
+    let mut out = CircuitData::new(Some(out_qubits), None, None, 0, Param::Float(0.)).unwrap();
```
Do we know what the upper bound on the size (as in the number of gates) is going to be here? If we do we should use `with_capacity()` instead of `new()` here to pre-allocate the internal vec.

ShellyGarion replied:
Here is an upper bound on the number of gates:
- according to the paper, the total number of CX gates without optimizations (a1 or a2) is `N = 9/16*4^n - 3/2*2^n`
- the total number of 1-qubit unitary gates is bounded by `2N`
- since we don't know which Euler decomposer will be chosen, we can bound the number of 1-qubit gates by `6N` 
- to conclude, the total capacity is bounded by `7N = 63/16*4^n - 21/2*2^n`

gadial replied:
Done in https://github.com/Qiskit/qiskit/pull/14797/commits/493f8602b7a6458505e0f925f1960089a01c9884

mtreinish commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,818 @@
+    let mat_check = a_block * b_block * c_block * Complex64::from(0.5);
+
+    const EPS: f64 = 1e-7;
+    abs_diff_eq!(mat, &mat_check, epsilon = EPS)
```
```suggestion
    abs_diff_eq!(mat, &mat_check, epsilon = 1e-7)
```

mtreinish reviewed (commented):
I think this is getting pretty close. Besides a bunch of inline suggestions. I had one more question.

mtreinish committed:
Apply suggestions from code review

mtreinish committed:
Run rustfmt

mtreinish commented on qiskit/synthesis/unitary/qsd.py:
```diff
@@ -106,6 +108,12 @@ def qs_decomposition(
+        decomposer_2q is None or isinstance(decomposer_2q, TwoQubitBasisDecomposer)
+    ):
+        array = np.asarray(mat, dtype=complex)
+        return QuantumCircuit._from_circuit_data(qsd.qs_decomposition(array, opt_a1, opt_a2))
```
You are missing the `decomposer_1q` and `decomposer_2q` arguments on the `qsd.qs_decomposition` call here. We've validated they are none or the correct type here, so we need to make sure that we pass the user specified arg to the rust code. Right now this is ignoring the user specified decomposers.

gadial replied:
Addressed in https://github.com/Qiskit/qiskit/pull/14797/commits/ce20bfbff428095042f62cea5bc9a0c40db3d6e1

alexanderivrii commented on crates/circuit/src/circuit_data.rs:
```diff
@@ -2310,6 +2310,41 @@ impl CircuitData {
         &self.qargs_interner
     }
 
+    /// Merge the `qargs` in a different [Interner] into this Circuit, remapping the qubits.
+    ///
+    /// This is useful for simplifying the direct mapping of [PackedInstruction]s from one circuit to
+    /// another, like when composing two circuits. See [Interner::merge_map_slice] for more
+    /// information on the mapping function.
+    ///
+    /// The input [InternedMap] is cleared of its previous entries by this method, and then we
+    /// re-use the allocation.
+    pub fn merge_qargs_using(
+        &mut self,
+        other: &Interner<[Qubit]>,
+        map_fn: impl FnMut(&Qubit) -> Option<Qubit>,
+        map: &mut InternedMap<[Qubit]>,
+    ) {
+        // 4 is an arbitrary guess for the amount of stack space to allocate for mapping the
+        // `qargs`, but it doesn't matter if it's too short because it'll safely spill to the heap.
+        self.qargs_interner
+            .merge_map_slice_using::<4>(other, map_fn, map);
+    }
+
+    /// Merge the `qargs` in a different [Interner] into this circuit, remapping the qubits.
+    ///
+    /// This is useful for simplifying the direct mapping of [PackedInstruction]s from one circuit to
+    /// another, like when composing two circuits. See [Interner::merge_map_slice] for more
+    /// information on the mapping function.
+    pub fn merge_qargs(
+        &mut self,
+        other: &Interner<[Qubit]>,
+        map_fn: impl FnMut(&Qubit) -> Option<Qubit>,
+    ) -> InternedMap<[Qubit]> {
+        let mut out = InternedMap::new();
+        self.merge_qargs_using(other, map_fn, &mut out);
+        out
+    }
```
Why does this show as new code -- isn't this a part of the main branch since at least #14313? I guess it would make sense to rebase this on top of main and fix a few merge conflicts that are bound to arise.

mtreinish replied:
These methods are being added to `CircuitData` they are equivalent to what's on `DAGCircuit` on main, but they don't exist on `CircuitData` on main. I debated spinning this out into a standalone PR, but the implementation is identical and figured it was small enough to add to this PR. But I can still separate it out if you'd prefer.

alexanderivrii replied:
Oh right (what was I thinking?). Either way is fine.

alexanderivrii commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,821 @@
+///
+/// \frac{22}{48} 4^n - \frac{3}{2} 2^n + \frac{5}{3}.
+///
+///
```
Here (and in some more places): `4^{n-2} - 1` and `\frac{22}{48} 4^n - \frac{3}{2} 2^n + \frac{5}{3}` are not indented consistently. I think the preferred option is not to indent this at all, see https://github.com/Qiskit/qiskit/pull/14988#discussion_r2324937256.

gadial replied:
I'm not 100% sure if not indenting would work; other places in the code use a single tab, so that's what I'll do.

alexanderivrii commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,821 @@
+///        `arXiv:0406176 [quant-ph] <https://arxiv.org/abs/quant-ph/0406176>`_
+///     2. Krol, Al-Ars, *Beyond Quantum Shannon: Circuit Construction for General
+///        n-Qubit Gates Based on Block ZXZ-Decomposition*,
+///        `arXiv:2403.13692 <https://arxiv.org/abs/2403.13692>`_
```
The above PR also describes the newly-preferred standard to format references.

alexanderivrii commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,821 @@
+/// * `one_qubit_decomposer`: Optional alternative one qubit euler basis to use for single qubit unitary decompositions. If not specified U is used.
+///
+/// # Returns
+///     Decomposed quantum circuit.
```
Does "Decomposed quantum circuit" need to be indented? I think the nightly clippy actually complains in such cases, but I have not checked.

alexanderivrii commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,821 @@
+/// # Arguments
+/// * `mat`: unitary matrix to decompose
+/// * `opt_a1`: whether to try optimization A.1
+/// * `opt_a2`: whether to try optimization A.2
```
From this description, it is not clear what `opt_a1` and `opt_a2` are, after all they are of type `Option<bool>`.

According to Shelly, the values `Some(true)` and `Some(false)` specify whether to enable or to disable the given optimization, and the value of `None` lets the algorithm to choose this automatically.

alexanderivrii commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,821 @@
+        for ctrl_index in 0..num_qubits {
+            let [um00, um11, um01, um10] = extract_multiplex_blocks(mat, ctrl_index);
+
+            if off_diagonals_are_zero(&um01, &um10, None) {
```
Imho, the check `if is_zero(&um01) && is_zero(&um10)` would be a bit clearer (no need to do this though).

gadial replied:
I agree; `off_diagonals_are_zero` is overly specific. I've changed this.

alexanderivrii commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,821 @@
+
+const EPS: f64 = 1e-10;
+
+enum VWType {
+    All,
+    OnlyV,
+    OnlyW,
+}
```
Can you add a comment explaining what this enum is used for?

gadial replied:
It's very algorithm-specific (and also I don't remember the details) but I've added that it's used in demultiplaxing.

alexanderivrii commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,821 @@
+    // the output circuit of the block ZXZ decomposition from [2]
+    let qr = (0..num_qubits).map(Qubit::new).collect::<Vec<_>>();
+    append(&mut out, left_circuit, &qr)?;
+    let _ = out.push_standard_gate(StandardGate::H, &[], &[Qubit((num_qubits - 1) as u32)]);
```
This should be a bit clearer (but please double-check the suggestion).
```suggestion
    out.push_standard_gate(StandardGate::H, &[], &[Qubit((num_qubits - 1) as u32)])?;
```
There is another similar line just below.

alexanderivrii commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,821 @@
+///
+/// where v and w are general unitaries determined from decomposition.
+///
+/// .. note::
```
This is pure Rust docs, so we should probably just write `/// Note:` or `/// # Note`.

alexanderivrii commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,821 @@
+    [(angle_1 + angle_2) / 2., (angle_1 - angle_2) / 2.]
+}
+
+fn append(circ: &mut CircuitData, new: CircuitData, qubit_map: &[Qubit]) -> PyResult<()> {
+    let new_qubits_map = circ.merge_qargs(new.qargs_interner(), |x| Some(qubit_map[x.index()]));
+    circ.add_global_phase(new.global_phase())?;
+    for inst in new.into_data_iter() {
+        let out_inst = PackedInstruction {
+            op: inst.op,
+            params: inst.params,
+            qubits: new_qubits_map[inst.qubits],
+            clbits: Default::default(),
+            label: inst.label,
+            #[cfg(feature = "cache_pygates")]
+            py_op: inst.py_op,
+        };
+        circ.push(out_inst)?;
+    }
+    Ok(())
+}
```
Oh nice, we should start using this more efficient code instead of `CircuitDataForSynthesis::compose` (not in this PR, of course).

alexanderivrii commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,821 @@
+        (VWType::OnlyV, true) => get_ucrz(num_qubits, &mut angles, false)?.reverse()?,
+        _ => get_ucrz(num_qubits, &mut angles, true)?,
+    };
+    // let multiplexed_rz = decompose_uc_rotation(&mut angles, StandardGate::RZ)?;
```
can this line be removed?

alexanderivrii commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,821 @@
+    append(
+        &mut out,
+        ucrz,
+        &[
+            std::slice::from_ref(layout.last().unwrap()),
+            &layout[..num_qubits - 1],
+        ]
+        .concat(),
```
This looks a bit complicated, but I understand that this avoids creating a new intermediate vector. But I am wondering if something like this would have been a bit more clear:
```rust
layout[..num_qubits - 1]
    .iter()
    .chain(std::iter::once(layout.last().unwrap()))
```

gadial replied:
This adds the `layout.last()` at the end, not the beginning; this can be fixed, but my main hurdle when trying to write this in the code was that the iterator needs to be `collected` into `&[Qubit]` and it complained about my attempts for doing it. I'm ok with keeping it as-is.

alexanderivrii commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,821 @@
+    let mut out = CircuitData::new(Some(out_qubits), None, None, 0, Param::Float(0.))?;
+    let q_target = Qubit(0);
+    let q_controls: Vec<Qubit> = (1..num_qubits).map(|i| Qubit(i as u32)).collect();
+    decompose_uc_rotations(angles, 0, angles.len(), false);
```
Ah, we did not have this function in Rust yet, this is why you had to implement this below?

gadial replied:
This is from Matthew's original code, so that's probably the reason. Can this be replaced by something else? Which one?

alexanderivrii commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,821 @@
+
+/// numpy's move_axis has the effect of pushing back the axis
+/// before the location we move into. there's no native ndarray equivalent
+fn move_axis<X>(array: ArrayD<X>, src: usize, dest: usize) -> ArrayD<X>
+where
+    X: Clone,
+{
+    let ndim = array.ndim();
+    let mut axes: Vec<usize> = (0..ndim).collect();
+    let axis = axes.remove(src);
+    axes.insert(dest, axis);
+    array.permuted_axes(axes)
+}
```
We don't have a similar function in `faer` or `nalgebra`, and we have to use `ndarray` for that?

(I was kind of hoping that in our quest of minimizing the number of different algebra packages, we don't need all 3 for implementing a specific algorithm).

gadial replied:
I added this in 0b7591fc6a0a4f33343b7ed7283958016acee6b6 to fix a bug caused by a smart code by Matthew which utilized numpy's amazing axes-shuffling skills (which, sadly, didn't really translate to `ndarry`). So it's a patch on a hack, and I'm sure we can do away with ndarry here.

However, I suggest we wait with this for a PR dedicated to our quest. Playing with axes can get you cut, so better not do it with an almost ready-to-merge PR.

alexanderivrii replied:
Haha! I wasn't suggesting any changes for this PR, I was asking for your opinion whether we can eventually get rid of `nadarray` (and only out of interest at that), your reply _"I'm sure we can do away with ndarry here"_ is good enough for me.

alexanderivrii commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,821 @@
+fn off_diagonals_are_zero(
+    um01: &DMatrix<Complex64>,
+    um10: &DMatrix<Complex64>,
+    atol: Option<f64>,
```
As I commented before, we don't really need this function as it's just two calls to something like `is_zero`. (Yeah, I know this was exactly the same in our Python code).

alexanderivrii commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,821 @@
+            (*idx, unitary.matrix(&[]).unwrap())
+        })
+        .collect();
+    for (ind1, ind2) in ind2q[0..ind2q.len() - 1].iter().zip(ind2q[1..].iter()) {
```
I think rust has a native method for that: https://doc.rust-lang.org/std/slice/struct.Windows.html. 
So you should be able to write something like `for &[ind1, ind2] in ind2q.windows(2) { ... }`.

gadial replied:
Nice!

alexanderivrii commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,821 @@
+        ),
+        VarsMode::Alike,
+    )?;
+    for (idx, inst) in circ.data().iter().enumerate() {
+        if let Some(new_circ) = diagonal_rollover.get(&idx) {
+            let block_index_map = circ.get_qargs(circ.data()[idx].qubits);
```
The code in this for-loop is the only code that I am still struggling to understand and that does seem to directly correspond to anything in the old Python code. Can you explain the intention of this?

alexanderivrii replied:
Update: this makes sense now (this replaces Python's 
`ccirc.data[ind2] = ccirc.data[ind2].replace(operation=qc3.to_gate())`.

alexanderivrii reviewed (commented):
Thank you all for the great work. Matthew was asking for someone to take another look at this PR, which I now did. The PR looks great, I only had some really minor questions and suggestions. From what I could see, the old Python implementation and the new Rust implementation match very closely (except for possibly some code in `apply_a2` that I am still trying to understand). I also think this warrants release notes.

gadial committed:
Better handling of default and python-supplied decomposers

gadial committed:
Create circuit with capacity based on heuristic upper bound

gadial committed:
Fixes according to PR review

alexanderivrii commented:
Thanks Gadi, LGTM! But let's wait for Matthew take another look as well.

mtreinish commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,880 @@
+            }
+        };
+    } else if dim == 4 {
+        if opt_a2 == Some(true) && depth > 0 {
+            let out_qubits = (0..num_qubits)
+                .map(|_| ShareableQubit::new_anonymous())
+                .collect::<Vec<_>>();
+            let mut out = CircuitData::new(Some(out_qubits), None, None, 0, Param::Float(0.))?;
+            let two_q_mat: Matrix4<Complex64> = Matrix4::from_fn(|i, j| mat[(i, j)]);
+            let packed_inst = PackedInstruction {
+                op: PackedOperation::from_unitary(Box::new(UnitaryGate {
+                    array: ArrayType::TwoQ(two_q_mat),
+                })),
+                qubits: out.add_qargs(&[Qubit(0), Qubit(1)]),
+                clbits: Default::default(),
+                params: None,
+                label: Some(Box::new("qsd2q".to_string())),
+                #[cfg(feature = "cache_pygates")]
+                py_op: OnceLock::new(),
+            };
+            out.push(packed_inst)?;
+            return Ok(out);
+        }
```
I know this was pre-existing in the python code so we should keep it, but this logic is weird if qsd is called with a 2q unitary and opt_a2 is `True`. Just another reason to follow up with deprecating these as user options.

mtreinish commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,880 @@
+    let mut zmat = DMatrix::<Complex64>::zeros(dim / 2, dim / 2);
+    for i in 0..dim / 2 {
+        zmat[(i, i)] = if i < dim / 4 {
+            Complex64::from(1.0)
```
```suggestion
            Complex64::ONE
```

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/14797/commits/a951c2b77efe924c14116a0feb01a37086e8ed2e

mtreinish commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,880 @@
+        zmat[(i, i)] = if i < dim / 4 {
+            Complex64::from(1.0)
+        } else {
+            Complex64::from(-1.0)
```
```suggestion
            -Complex64::ONE
```

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/14797/commits/a951c2b77efe924c14116a0feb01a37086e8ed2e

mtreinish commented on crates/synthesis/src/qsd.rs:
```diff
@@ -0,0 +1,880 @@
+            Complex64::from(1.0)
+        } else {
+            Complex64::from(-1.0)
+        };
```
Should this be inside the condition on `opt_a1_val`? It's not used anywhere else I could see.

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/14797/commits/a951c2b77efe924c14116a0feb01a37086e8ed2e

mtreinish reviewed (commented):
A few super small comments and this is still missing a release note. But I think after thats addressed this should be good to merge

mtreinish committed:
Use constants for zmat and only construct when needed

mtreinish committed:
Fix future clippy error

mtreinish committed:
Add release note

mtreinish reviewed (approved):
This LGTM now, thanks for all the work on this everyone. Between @alexanderivrii's earlier approval and mine here I think that is sufficient review even though both @alexanderivrii and I contributed fairly significantly to this PR we didn't overlap and I think we've had enough eyes on this for the past couple of weeks that this is good to merge.
